vision.md

```diff
@@ -206,26 +206,188 @@ The framework includes modules for cloud connectivity, data mapping, device
+      solutions, using (networking)interfaces like for example MQTT to attach
+      their own infrastructure to thin-edge.io, or using the final
+      deployment-ready binary to solve their problems).
+1. Introspective functionality to enable quality assurance as well as support of
+   deployments
+    - Knowing what happens inside a system is crucial to be able to quickly
+      identify misbehaving components. Users are be able to access an overview
+      of the communication of the different components
+1. Ability to attach data sources and data sinks to thin-edge.io via different
+   mechanisms, either by implementing connectors directly in the framework using
+   rust code, attaching them via some networking mechanism (e.g. MQTT) or other
+   mechanisms the framework provides
+    1. The framework ships with connectivity functionality for the most common
+       connectivity requirements
+    1. Configuration of connectivity functionality is exposed to the user
+    1. Connectivity over the same mechanism but with different configuration
+       settings is easily doable for the user (e.g. using MQTT to connect to
+       different brokers with completely different settings)
+1. Easy deployment and little friction when using thin-edge.io as a framework to
+   develop their own solutions based on it
+1. Trust in the user to do the right thing with the tools provided, respect
+   their decisions and guide them towards a maintainable solution by design
+    - Users are assumed to be experts in their domain and should be given the
+      tools they need to solve the problems they have. Unnecessarily burdening
+      them with abstractions only causes frustration
+1. High configurability and tweakability with high trust in custom changes
+   without having to redefine the world on small changes
+1. Allow precise configuration inside common solutions shipped by the framework
+    - Users are be able to change a small detail of their deployment without
+      having to then define unrelated elements
+    - As an example: Users wanting to set an SSL certificate should not need to
+      worry about also defining their DNS setup.
+1. Reproducability of a configuration
+1. Security
+    1. Operational Security
+        1. Components of the framework are encapsulated and do not influence
+           each others execution
+        1. The system can recover from bad states in a safe way
+        1. Untrusted input is handled in a way that does not influence execution
+           in a bad way (e.g. crashes the system)
+    1. Information Security
+        1. Data access is reglemented
+        1. Data origin is verified and potentially encrypted
+    1. Code Audit(ability)
+        1. Functionality can be audited easily because it is encapsulated (see
+           above)
+1. Generic solutions over specific solutions
+    - Problems are solved in a way that users can define the details as best
+      suited
```

Comment:
Can you provide an example of a generic solution vs a specific solution?

Reply:
A specific solution is to make a component that makes an HTTP call to a specific URL.

A generic solution would send a `HTTPRequest` message to for example a `HTTP` component.



Sometimes the specific solution is required, but the generic one is preferred.

Reply:
> Data access is reglemented
2 Data origin is verified and potentially encrypted
Functionality can be audited easily because it is encapsulated (see
above)
--> what is the reason for those requirements ? Seems to me like we are preparing thin edge to handle mission-critical or high security systems, if this is the case from a business point we should discuss which IoT use cases we are going after. We are not intending to serve those to my understanding and need to verify what drives those NFR suggestions

I'm aligned with @AndrejS321. These requirements are specific to some particular use-cases - healthcare being a notable example. However, these particular use-cases matter. So we might add a requirement to be able to add such constraints (say be replacing a light MQTT transport layer by an encrypted/auditable transport layer).